File: foodmate-application/src/main/java/com/foodmate/application/food/service/CompositeDishService.java
package com.foodmate.application.food.service;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;

/** 用户复合菜基础用例；不提供公共菜谱发布和烹饪步骤管理。 */
public interface CompositeDishService {
    CompositeDishView create(long userId, CreateCommand command);

    CompositeDishView get(long userId, long compositeDishId);

    List<CompositeDishView> list(long userId);

    CompositeDishView update(
            long userId, long compositeDishId, long revision, UpdateCommand command);

    void delete(long userId, long compositeDishId, long revision, String idempotencyKey);

    record CreateCommand(
            String dishName,
            BigDecimal totalServings,
            List<ComponentCommand> components,
            String idempotencyKey) {
        public CreateCommand(
                String dishName, BigDecimal totalServings, List<ComponentCommand> components) {
            this(dishName, totalServings, components, null);
        }

        public CreateCommand {
            components = components == null ? List.of() : List.copyOf(components);
        }
    }

    record UpdateCommand(
            String dishName,
            BigDecimal totalServings,
            List<ComponentCommand> components,
            String idempotencyKey) {
        public UpdateCommand {
            components = components == null ? List.of() : List.copyOf(components);
        }
    }

    record ComponentCommand(long nutritionFoodId, String rawName, BigDecimal amount, String unit) {}

    record CompositeDishView(
            long compositeDishId,
            String dishName,
            BigDecimal totalServings,
            BigDecimal caloriesKcalPerServing,
            BigDecimal proteinGPerServing,
            BigDecimal fatGPerServing,
            BigDecimal carbsGPerServing,
            String nutritionSource,
            long revision,
            boolean deleted,
            Instant createdAt,
            Instant updatedAt,
            List<ComponentView> components) {}

    record ComponentView(
            long itemId,
            int itemOrder,
            long nutritionFoodId,
            String rawName,
            BigDecimal amount,
            String unit,
            BigDecimal normalizedAmount,
            String normalizedUnit,
            BigDecimal caloriesKcal,
            BigDecimal proteinG,
            BigDecimal fatG,
            BigDecimal carbsG) {}
}

File: foodmate-ui/src/services/compositeDishService.ts
import { apiRequest } from './apiClient';

export type CompositeDishComponent = {
  item_id: string;
  item_order: number;
  nutrition_food_id: string;
  raw_name: string;
  amount: number | string;
  unit: string;
  normalized_amount: number | string;
  normalized_unit: string;
  calories_kcal: number | string;
  protein_g: number | string;
  fat_g: number | string;
  carbs_g: number | string;
};

export type CompositeDish = {
  composite_dish_id: string;
  dish_name: string;
  total_servings: number | string;
  calories_kcal_per_serving: number | string;
  protein_g_per_serving: number | string;
  fat_g_per_serving: number | string;
  carbs_g_per_serving: number | string;
  nutrition_source: string;
  revision: number;
  deleted: boolean;
  created_at: string;
  updated_at: string;
  components: CompositeDishComponent[];
};

export type CompositeDishWriteRequest = {
  dish_name: string;
  total_servings: number;
  components: Array<{ nutrition_food_id: number; raw_name: string; amount: number; unit: string }>;
};

export async function loadCompositeDishes(signal?: AbortSignal): Promise<CompositeDish[]> {
  return apiRequest<CompositeDish[]>('/api/composite-dishes', signal ? { signal } : undefined);
}

export async function loadCompositeDish(compositeDishId: string, signal?: AbortSignal): Promise<CompositeDish> {
  return apiRequest<CompositeDish>(
    `/api/composite-dishes/${encodeURIComponent(compositeDishId)}`,
    signal ? { signal } : undefined,
  );
}

export async function createCompositeDish(
  request: CompositeDishWriteRequest,
  signal?: AbortSignal,
): Promise<CompositeDish> {
  return apiRequest<CompositeDish>('/api/composite-dishes', {
    method: 'POST',
    headers: { 'Idempotency-Key': idempotencyKey('composite-dish-create') },
    body: JSON.stringify(request),
    signal,
  });
}

export async function updateCompositeDish(
  compositeDishId: string,
  revision: number,
  request: CompositeDishWriteRequest,
  signal?: AbortSignal,
): Promise<CompositeDish> {
  return apiRequest<CompositeDish>(
    `/api/composite-dishes/${encodeURIComponent(compositeDishId)}?revision=${revision}`,
    {
      method: 'PATCH',
      headers: { 'Idempotency-Key': idempotencyKey('composite-dish-update') },
      body: JSON.stringify(request),
      signal,
    },
  );
}

export async function deleteCompositeDish(
  compositeDishId: string,
  revision: number,
  signal?: AbortSignal,
): Promise<void> {
  await apiRequest<void>(`/api/composite-dishes/${encodeURIComponent(compositeDishId)}?revision=${revision}`, {
    method: 'DELETE',
    headers: { 'Idempotency-Key': idempotencyKey('composite-dish-delete') },
    signal,
  });
}

function idempotencyKey(prefix: string) {
  const suffix = globalThis.crypto?.randomUUID?.() ?? `${Date.now()}-${Math.random().toString(16).slice(2)}`;
  return `${prefix}-${suffix}`;
}
